singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,426 @@
+        } else {
+          int count = eventTriedCount.get(event.getLoggingAuditHeaders());
+          if (count >= numOfPartitionsToTrySending) {
+            LOG.debug("Failed to send audit event after trying {} partitions. Drop this event.",
+                count);
+            OpenTsdbMetricConverter
+                .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_EVENTS_DROPPED, 1,
+                    "host=" + host, "stage=" + stage.toString(),
+                    "logName=" + event.getLoggingAuditHeaders().getLogName());
+            eventTriedCount.remove(event.getLoggingAuditHeaders());
+          } else {
+            eventTriedCount.put(event.getLoggingAuditHeaders(), count + 1);
+            boolean success = false;
+            try {
+              success = queue.offerFirst(event, 3, TimeUnit.SECONDS);
+              if (!success) {
+                LOG.debug(
+                    "Failed to enqueue LoggingAuditEvent at head of the queue when executing "
+                        + "producer send callback. Drop this event.");
+                eventTriedCount.remove(event.getLoggingAuditHeaders());
+              }
+            } catch (InterruptedException ex) {
+              LOG.debug(
+                  "Enqueuing LoggingAuditEvent at head of the queue was interrupted in callback. "
+                      + "Drop this event");
+              eventTriedCount.remove(event.getLoggingAuditHeaders());
+            }
+          }
+        }
+      } catch (Throwable t) {
+        LOG.warn("Exception throws in the callback. Drop this event {}", event, t);
+        OpenTsdbMetricConverter
+            .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_CALLBACK_EXCEPTION, 1,
+                "host=" + host, "stage=" + stage.toString(), "topic=" + topic);
+      }
+    }
+  }
+
+  /**
+   *  start the executing thread and let the Sender run.
+   */
+  public synchronized void start() {
+    if (this.thread == null) {
+      thread = new Thread(this);
+      thread.setDaemon(true);
+      thread.setName(name);
+      thread.start();
+      LOG.warn(
+          "[{}] created and started [{}] to let it dequeue LoggingAuditEvents and send to Kafka.",
+          Thread.currentThread().getName(), name);
+    }
+  }
+
+  /**
+   *  reserve some time (300 seconds at most)to let AuditEventKafkaSender to send out
+   *  LoggingAuditEvent in the queue and gracefully stop AuditEventKafkaSender.
+   */
+  public synchronized void stop() {
+    LOG.warn(
+        "[{}] waits up to {} seconds to let [{}] send out LoggingAuditEvents left in the queue if"
+            + " any.",
+        Thread.currentThread().getName(), stopGracePeriodInSeconds, name);
+    int i = 0;
+    int numOfRounds = stopGracePeriodInSeconds / callingThreadSleepInSeconds;
+    while (queue.size() > 0 && this.thread != null && thread.isAlive() && i < numOfRounds) {
+      i += 1;
+      try {
+        Thread.sleep(callingThreadSleepInSeconds * 1000);
```

Comment:
with this `Thread.sleep(....)`,  the process will wait for 5 minutes before it can exit. this can be problematic for some services.  shall we reduce the default waiting time to 10 seconds?   if there is message delivery failure within 10 seconds, in most of the cases having 300 seconds waiting time will not be helpful as the service will not be able to recovery within 5 minutes, given that we have already had retries on sending messages to different partitions.

Reply:
5 mins is the maximum amount of time to wait.  If there is no events in the queue or a small number of events that quickly get delivered in a short period (say 90 seconds), the wait will stop earlier.  Reserving some time during the shutdown is just trying the best to avoid audit events loss in order to give more accurate auditing results later. 

Since different systems (services using thrift logger, Singer, Merced, etc) can tolerate different amount of time for grace shutdown,  I will make this a configuration property with default 5 mins.  If certain systems don't want that long, a smaller value should be specified.